### ISSUE TITLE
CI failure messages over #buildkite-main are not actionnable enough

### ISSUE THREAD

**Issue by jhchabran** [sourcegraph/sourcegraph#37613]
Right now, when a build fails, the standard message failure message is posted in #buildkite-main, mentioning its author. 

The only thing it achieves is notifying the author that something went wrong. So the channel is a long list of failures, with some of them being threads where teammates are debating how to fix the problem or address the flake. 

The flow is usually that someone will take a look at the failure, check what caused it, and will eventually post it in the chan if there is reason to think it could be fixed shortly. 

Therefore, unless you're actively monitoring or getting pinged, you'll never see that your flaking tests are affecting others, which when a test suite stays flaky for too long, usually end up being ignored. 

By adding more information to the failure messages, we can put the source of failures in front of us, which will create more engagement to fix problems, because you can't ignore that _your test suite is interrupting your colleague_. 

This is not the ultimate solution to creating more engagement around failures, but one of the measures we can take to spark more initiative and reactiveness from everyone.

**Comment by jhchabran** [comment]
Ran some quick discovery: 

- Using the plugin https://github.com/sourcegraph/step-slack-notify-buildkite-plugin is not going to work, because this is a per step notification and that's going to be very very noisy. 
- There are no hooks for a completed build, as only the Buildkite servers are aware of the final state of a build. We could probably poll that at the end of each step, that would be horribly wrong. 
- The best approach would be to use a webhook, which is was the Buildkite support advised to use when I reached out to them a while ago. 
  - That would required building and deploying a small service to receive the hooks, that's quite straightforward to do, and could be a good candidate to learn more about Go. 

We have to ask ourselves if that's still worth it, but the added visibility on the build logs may be quite welcomed, as it will increase failures actionability. But it also open the door to leverage Slack integrations, to create issues with a Slack button for example, to say that something is flaky. 

cc @burmudar you probably want to take a look at this!

TL;DR: it's slightly more complicated that we thought, it's probably 1d of work at most.

**Comment by burmudar** [comment]
So initially I went the `pre-exit` hook route but ended up going with the [webhook route](https://buildkite.com/docs/apis/webhooks) which, so far, is really promising. I've setup a webhook for two events `job.finished` and `build.finished`.

Job finished allows us to keep track of invidual job failures and build finished allows us to know when a build has finished. Once the build is finished we trigger the building of the notification for failed steps, and possibly other actions!

**Comment by valerybugakov** [comment]
Excited about this improvement! This issue should further decrease the number of ignored flakes if we add a message like this: "PLEASE DO NOT IGNORE FLAKES and follow respective guidelines [link] to disable them". 🙂

**Comment by valerybugakov** [comment]
Two valuable pieces of documentation to link to the failure message would be:

1. [CI flakes](https://docs.sourcegraph.com/dev/background-information/ci#flakes)
2. [Assessing flaky client steps](https://docs.sourcegraph.com/dev/how-to/testing#assessing-flaky-client-steps)

[Slack thread for context](https://sourcegraph.slack.com/archives/C01N83PS4TU/p1658203376205619).
